### PR TITLE
Feature map generation basic

### LINKED ISSUE #79: Advanced maze generation
As developers we want to create an algorithm that generates random traversable mazes for us, so that our game won’t get repetitive.

Subtasks:
- Explore possibilities
- Implement algorithm
- Make loading and saving rooms possible
- Combine the features above

### LINKED ISSUE #80: Advanced maze generation - Explore
Explore possibilities for generating mazes.

Subtask of #79 - Advanced maze generation

### LINKED ISSUE #81: Advanced maze generation - Algorithm
Create an algorithm for generating mazes.

Subtask of #79

### LINKED ISSUE #83: Advanced maze generation - Combining
Combine all the parts to create full maze generation.
This uses subtasks #80, #81 and #82.

Subtask of #79 - Advanced maze generation

### PR DESCRIPTION
**Pull request by Geweldig**
This implements the random level generation.
This PR contains major changes to the following classes:
-  MazeTile, 97% coverage
-  Level, 100% coverage
-  RandomLevelFactory, 92% coverage

This PR contains the following new classes:
-  Size, 100% coverage
-  GeneratorRoom, 100% coverage

There are no PMD, FindBugs, or Checkstyle warnings (after the changes made in the master will be merged).
Two of the tests in RandomLevelFactory are "bad" tests as they only check a value is returned, but don't check the entire process. This is adressed upon in their respective comments.
As (most) of the methods they call have been fully tested it's safe to keep these tests until we have decided upon our final parameters.

This closes #79, closes #80, closes #81, closes #83

### CODE REVIEW

**Comment by Geweldig** [comment]
Figured I would upload a screenshot of it in action.
In action

File: src/main/java/nl/tudelft/contextproject/model/level/RandomLevelFactory.java [hunk 1]
```diff
@@ -1,108 +1,340 @@
...
+	@SneakyThrows(URISyntaxException.class)
+	protected List<Size> loadRooms() {
+		File[] files = new File(RandomLevelFactory.class.getResource("/rooms").toURI()).listFiles();
+
+		/**
+		 * If the folder is not a folder, we throw an IllegalStateException exception.
+		 */
+		if (files == null) {
+			throw new IllegalStateException("There are no rooms.");
+		}
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/level/RandomLevelFactory.java:119]: Comment says not a folder and error message there are no rooms in the folder

File: src/main/java/nl/tudelft/contextproject/model/level/RandomLevelFactory.java [hunk 2]
```diff
@@ -1,108 +1,340 @@
...
+	 * Method to create a list of {@link GeneratorRoom}.
+	 *
+	 * @return
+	 * 			the created generatorrooms
+     */
+	protected List<GeneratorRoom> create() {
+		List<GeneratorRoom> rooms = new ArrayList<>();
+		List<Size> sizes = loadRooms();
+		if (!allowDuplicates && sizes.size() < amount) {
+			throw new IllegalArgumentException("You are requesting more rooms than there are.");
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/level/RandomLevelFactory.java]: Maybe add sizes.size() to this message to give the user a maximum value to choose from.

File: src/main/java/nl/tudelft/contextproject/model/level/RandomLevelFactory.java (hunk 1 above)
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/level/RandomLevelFactory.java:119]: My thinking was that if the folder isn't there there are no rooms. This is indeed extremely vague, and will be fixed.

**Comment by ChielBruin** [comment]
Made two miror (nitpicky) comments, but the code looks good and the result is great!

**Comment by Geweldig** [comment]
I have addressed both of your issues as they do indeed make the exceptions easier to understand. 
Could you review these changes at your convenience?

**Comment by ChielBruin** [comment]
Now it's good to merge for me!

**Comment by Diocruel** [comment]
Looks great to me. Merging.

**Pull request merged**